flight_sql/accessors/timestamp_array_accessor.cc

```diff
@@ -32,23 +34,36 @@ int64_t GetConversionToSecondsDivisor(TimeUnit::type unit) {
   return divisor;
 }
 
-uint32_t CalculateFraction(TimeUnit::type unit, uint64_t units_since_epoch) {
+uint32_t CalculateFraction(TimeUnit::type unit, int64_t units_since_epoch) {
   // Convert the given remainder and time unit to nanoseconds
   // since the fraction field on TIMESTAMP_STRUCT is in nanoseconds.
   switch (unit) {
```

Comment:
Is it possible to eliminate this switch entirely? If we make the shifting calculation its own function (described in my other comment) then we are pretty much left with just the code to scale the fraction to nanoseconds.

Can that be expressed like units_since_epoch % (NANOS_TO_SECONDS_DIVISOR / GetSecondsDivisor()). The right-hand of the mod operator should get optimized out into a constant and the switch would be eliminated. (might have gotten the division order backwards).

Reply:
More or less yet, and likewise with the return statements where I wish (easy enough to fix) the 10**n magic numbers were derived as constexpr.  However we're left with the shift-to-force-rounding-direction, which is in the opposite direction in different places here (floor div vs ceiling div) so I would either be duplicating the function to handle that detail both ways OR negating the operands to get the same effect which is pretty hacky (and buggy for unchecked signed ints).